oci/casext/gc_test.go

```diff
@@ -175,3 +177,151 @@ func TestGCWithNonEmptyIndex(t *testing.T) {
+	return func(digest digest.Digest) (bool, error) {
+		if digest != expectedDigest {
+			t.Errorf("got incorrect digest to gc policy callback: expected %v, got %v", expectedDigest, digest)
+		}
+		return false, nil
+	}
+}
+
+func errFunc(digest digest.Digest) (bool, error) {
+	return false, fmt.Errorf("err policy")
+}
+
+func TestGCWithPolicy(t *testing.T) {
+	ctx := context.Background()
+
+	root, err := ioutil.TempDir("", "umoci-TestEngineReference")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(root)
+
+	image := filepath.Join(root, "image")
+	if err := dir.Create(image); err != nil {
+		t.Fatalf("unexpected error creating image: %+v", err)
+	}
+
+	engine, err := dir.Open(image)
+	if err != nil {
+		t.Fatalf("unexpected error opening image: %+v", err)
+	}
+	engineExt := NewEngine(engine)
+	defer engine.Close()
+
+	// build a orphan blob that should be GC'ed
+	content := "this is a orphan blob"
+	br := strings.NewReader(content)
+	odigest, size, err := engine.PutBlob(ctx, br)
+	if err != nil {
+		t.Fatalf("error writing blob: %+v", err)
+	}
+	if size != int64(len(content)) {
+		t.Fatalf("partially written blob")
+	}
+
+	// build a blob, manifest, index that will survive GC
+	content = "this is a test blob"
+	br = strings.NewReader(content)
+	digest, size, err := engine.PutBlob(ctx, br)
+	if err != nil {
+		t.Fatalf("error writing blob: %+v", err)
+	}
+	if size != int64(len(content)) {
+		t.Fatalf("partially written blob")
+	}
+
+	m := ispec.Manifest{
+		Versioned: imeta.Versioned{
+			SchemaVersion: 2,
+		},
+		Config: ispec.Descriptor{
+			MediaType: ispec.MediaTypeImageLayer,
+			Digest:    digest,
+			Size:      size,
+		},
+		Layers: []ispec.Descriptor{
+			{
+				MediaType: ispec.MediaTypeImageLayer,
+				Digest:    digest,
+				Size:      size,
+			},
+		},
+	}
+	data, err := json.Marshal(&m)
+	if err != nil {
+		t.Fatalf("error marshaling json: %+v", err)
+	}
+	mr := bytes.NewReader(data)
+	digest, size, err = engine.PutBlob(ctx, mr)
+	if err != nil {
+		t.Fatalf("error writing blob: %+v", err)
+	}
+	if size != int64(len(data)) {
+		t.Fatalf("partially written blob")
+	}
+
+	idx := ispec.Index{
+		Versioned: imeta.Versioned{
+			SchemaVersion: 2,
+		},
+		Manifests: []ispec.Descriptor{
+			{
+				MediaType: ispec.MediaTypeImageManifest,
+				Digest:    digest,
+				Size:      size,
+			},
+		},
+	}
+	if err := engine.PutIndex(ctx, idx); err != nil {
+		t.Fatalf("error writing index: %+v", err)
+	}
+
+	err = engineExt.GC(ctx, errFunc)
+	// expect this to fail
+	if err == nil {
+		t.Fatalf("GC failed: %+v", err)
+	}
+
+	err = engineExt.GC(ctx, gcSkipFunc(t, odigest))
+	// expect this to succeed but not perform GC
+	if err != nil {
+		t.Fatalf("GC failed: %+v", err)
+	}
+	b, err := engine.ListBlobs(ctx)
+	if err != nil {
+		t.Fatalf("unable to list blobs: %+v", err)
+	}
+	if len(b) == 1 {
+		t.Fatalf("didn't expect single-entry blob list after skip GC policy")
+	}
```

Comment:
This should check `if len(b) != 3` (`== 1` should've been `== 2` but `!= 3` is less fragile and more accurately describes the behaviour we're testing) and have an updated error message.